File: android/app/src/main/kotlin/com/stockbook/app/AppRouter.kt
package com.stockbook.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.stockbook.core.model.Bill
import com.stockbook.core.model.CreditNote
import com.stockbook.core.model.Payment
import com.stockbook.core.model.Customer
import com.stockbook.core.model.Expense
import com.stockbook.core.model.Product
import com.stockbook.core.model.Purchase
import com.stockbook.core.model.Supplier
import com.stockbook.core.model.SupplierPayment
import com.stockbook.core.text.AppTab

/**
 * Navigation state.
 *
 * The handoff describes a flat router plus overlays that can sit above any
 * screen, so that is what this is rather than a `NavHost` per tab. There is no
 * drill-down anywhere in this app: every detail view is a bottom sheet or a
 * full-screen overlay, which is also why nothing here needs a back stack.
 */
class AppRouter {
    var tab by mutableStateOf(AppTab.TODAY)

    /** Settings is reached from the Today gear, not from the tab bar. */
    var showingSettings by mutableStateOf(false)

    /** The export/import handoff, one level in from Settings. */
    var showingBackup by mutableStateOf(false)

    /** Null closed; a product means edit, `NEW_PRODUCT` means create. */
    var productEditor by mutableStateOf<Product?>(null)
    var creatingProduct by mutableStateOf(false)

    /**
     * The expense sheet: the expense being corrected, or [creatingExpense] for a
     * new one. The same pair the product editor uses, for the same reason — the
     * sheet needs to know which of the two it is without a nullable meaning two
     * different things.
     */
    var expenseEditor by mutableStateOf<Expense?>(null)
    var creatingExpense by mutableStateOf(false)

    /**
     * The stock sheet, opened from a product: it can count that product's shelf
     * or file a supplier's bill against it.
     */
    var addStock by mutableStateOf<Product?>(null)

    /**
     * Set by the Items header's Delivery button: the same sheet with no product
     * named, which is a supplier bill and nothing else.
     *
     * Naming the product used to be the *first* question a delivery asked, in a
     * sheet of its own. It is optional now — a bill for a mixed load names
     * nothing and still owes money — so the question moved inside, where it can
     * be left alone.
     */
    var recordingDelivery by mutableStateOf(false)

    /** A delivery opened from the book, the way a bill is opened from history. */
    var purchaseDetail by mutableStateOf<Purchase?>(null)

    /**
     * The delivery being corrected, or null when the stock sheet is recording a
     * new one. A third door into that same sheet, deliberately: a correction typed
     * on a screen of its own is a screen that drifts away from the one the
     * delivery was entered on.
     */
    var editingPurchase by mutableStateOf<Purchase?>(null)
        private set

    /**
     * The two Today banners, opened into a list of everybody behind them.
     *
     * A banner saying "3 customers still owe" is a fact the owner can do nothing
     * with; these turn it into the names, and each name into the payment sheet.
     */
    /**
     * Whether Sell is showing the product picker rather than the bill form.
     *
     * Screen-local by rights, and here because the shell has to know: the picker
     * carries its own bottom bar, and the tab bar underneath it would be a second
     * one stacked on the first. Reset when Sell leaves the screen, so coming back
     * lands on the form rather than wherever the last visit ended.
     */
    var pickingProducts by mutableStateOf(false)

    var showingDebtors by mutableStateOf(false)
    var showingCreditors by mutableStateOf(false)

    /**
     * Which day the day summary is showing, or null when it is closed.
     *
     * The day itself rather than a flag, because the sheet steps between days
     * and the one it is on has to survive a recomposition. Opened on today from
     * the date at the top of Home.
     */
    var dayInView by mutableStateOf<java.time.Instant?>(null)

    /** The receipt, shown full-screen after a bill is saved. */
    var receipt by mutableStateOf<Bill?>(null)

    /**
     * A bill opened from history. Distinct from [receipt]: that one confirms
     * something that just happened, this one is a document being looked up.
     */
    var billDetail by mutableStateOf<Bill?>(null)

    /**
     * The bill being corrected, or null when Sell is writing a new one.
     *
     * Sell is the screen every bill in this shop was typed on, so it is the screen
     * a correction is typed on too — this is what tells it which of the two it is
     * doing, and which bill `updateBill` is being handed.
     */
    var editingBill by mutableStateOf<Bill?>(null)
        private set

    /**
     * Which tab Edit was tapped from, so finishing a correction goes back there.
     *
     * Without it the owner lands on a blank bill form after saving, which reads as
     * the app inviting them to write another one.
     */
    private var tabBeforeEditing = AppTab.TODAY

    /**
     * The customer editor sheet. Null closed; a customer means correct, and
     * [creatingCustomer] means create.
     */
    var customerEditor by mutableStateOf<Customer?>(null)
    var creatingCustomer by mutableStateOf(false)

    /** The record-a-payment sheet, for one customer. */
    var paymentFor by mutableStateOf<Customer?>(null)

    /**
     * The credit-note sheet, for one customer. Its sibling below carries the
     * note being corrected — nil for a new one, exactly as the bill editor does.
     */
    var creditNoteFor by mutableStateOf<Customer?>(null)
    var editingCreditNote by mutableStateOf<CreditNote?>(null)

    /**
     * The payment being corrected, if one is. Held beside [paymentFor] rather
     * than replacing it: the sheet needs the customer either way, because it
     * shows what will still be owed once the correction is saved.
     */
    var editingPayment by mutableStateOf<Payment?>(null)

    /**
     * A customer's statement, full screen. Held as a **key** rather than a
     * [Customer], because recording a payment while it is open changes every
     * derived figure on it — the screen has to re-read the customer, not show a
     * copy taken when it opened.
     */
    var statementFor by mutableStateOf<String?>(null)

    /**
     * One party's screen, full width: a customer key, or a supplier key with
     * [partyIsSupplier] set.
     *
     * A key rather than the `Customer` itself, for the same reason [statementFor]
     * is a key — taking a payment while the screen is open changes every figure on
     * it, so the screen has to re-read the person rather than show a copy taken
     * when it opened.
     *
     * One field and a flag rather than two fields, unlike the statements above:
     * the two statements really are two presentations because only one of them
     * can be open at a time *per side*, whereas a party screen is one place with a
     * direction. Both spellings work; this is the one that stops the two from
     * being open at once.
     */
    var partyFor by mutableStateOf<String?>(null)
    var partyIsSupplier by mutableStateOf(false)

    fun openCustomerScreen(key: String) {
        partyIsSupplier = false
        partyFor = key
    }

    fun openSupplierScreen(key: String) {
        partyIsSupplier = true
        partyFor = key
    }

    /**
     * The supplier editor sheet. The customer editor's mirror, kept as its own
     * pair of fields rather than one editor with a direction on it: the two
     * sheets say different words and gate on different figures.
     */
    var supplierEditor by mutableStateOf<Supplier?>(null)
    var creatingSupplier by mutableStateOf(false)

    /** The pay-a-supplier sheet. */
    var supplierPaymentFor by mutableStateOf<Supplier?>(null)

    /** The money-out twin of [editingPayment]. */
    var editingSupplierPayment by mutableStateOf<SupplierPayment?>(null)

    /**
     * A supplier's statement, full screen — a key for the same reason
     * [statementFor] is one, and a separate field so the screen knows which side
     * of the book it is drawing without being told twice.
     */
    var supplierStatementFor by mutableStateOf<String?>(null)

    fun openNewSupplier() {
        supplierEditor = null
        creatingSupplier = true
    }

    fun openSupplier(supplier: Supplier) {
        creatingSupplier = false
        supplierEditor = supplier
    }

    fun openSupplierStatement(supplier: Supplier) {
        supplierStatementFor = supplier.key
    }

    fun closeSupplierEditor() {
        supplierEditor = null
        creatingSupplier = false
    }

    fun openNewCustomer() {
        customerEditor = null
        creatingCustomer = true
    }

    fun openCustomer(customer: Customer) {
        creatingCustomer = false
        customerEditor = customer
    }

    fun openStatement(customer: Customer) {
        statementFor = customer.key
    }

    fun closeCustomerEditor() {
        customerEditor = null
        creatingCustomer = false
    }

    fun openNewProduct() {
        productEditor = null
        creatingProduct = true
    }

    fun openNewExpense() {
        expenseEditor = null
        creatingExpense = true
    }

    fun openExpense(expense: Expense) {
        creatingExpense = false
        expenseEditor = expense
    }

    fun closeExpense() {
        expenseEditor = null
        creatingExpense = false
    }

    fun openProduct(product: Product) {
        creatingProduct = false
        productEditor = product
    }

    fun openAddStock(product: Product) {
        productEditor = null
        creatingProduct = false
        addStock = product
    }

    /** Closes the stock sheet from any of the three ways it can be open. */
    fun closeAddStock() {
        addStock = null
        recordingDelivery = false
        editingPurchase = null
    }

    fun openBill(bill: Bill) {
        billDetail = bill
    }

    /**
     * Takes a bill out of its sheet and onto the form it was written on. The cart
     * is filled by the caller, which is the only thing here that knows one exists.
     */
    fun editBill(bill: Bill) {
        billDetail = null
        tabBeforeEditing = tab
        editingBill = bill
        // The picker is where the last bill left it, and a correction that opened
        // on a product list would hide the figures it came to change.
        pickingProducts = false
        tab = AppTab.SELL
    }

    /** Leaves a correction — saved or abandoned — and puts the owner back. */
    fun closeBillEditing() {
        if (editingBill == null) return
        editingBill = null
        pickingProducts = false
        tab = tabBeforeEditing
    }

    /** Opens a delivery on the sheet it was entered on, in place of its detail. */
    fun editPurchase(purchase: Purchase) {
        purchaseDetail = null
        editingPurchase = purchase
    }

    fun startBill() {
        tab = AppTab.SELL
    }

    fun closeOverlays() {
        expenseEditor = null
        creatingExpense = false
        partyFor = null
        productEditor = null
        creatingProduct = false
        addStock = null
        recordingDelivery = false
        purchaseDetail = null
        editingPurchase = null
        showingDebtors = false
        showingCreditors = false
        dayInView = null
        receipt = null
        billDetail = null
        editingBill = null
        showingBackup = false
        customerEditor = null
        creatingCustomer = false
        paymentFor = null
        editingPayment = null
        creditNoteFor = null
        editingCreditNote = null
        statementFor = null
        supplierEditor = null
        creatingSupplier = false
        supplierPaymentFor = null
        editingSupplierPayment = null
        supplierStatementFor = null
    }
}

File: ios/Stockbook/App/AppRouter.swift
import Foundation
import Observation

/// Navigation state.
///
/// The handoff describes a flat router — `setup0 | setup1 | setup2 | home |
/// items | sell | bills | settings`, plus three overlays that can sit above any
/// screen — so that is what this is, rather than a `NavigationStack` per tab.
/// There is no drill-down anywhere in the app: every detail view is a bottom
/// sheet or a full-screen overlay.
///
/// Setup is not held here. Whether it has run is *persisted* state
/// (`ShopSettings.setupCompleted`), and `RootView` branches on it.
@Observable
final class AppRouter {

    var tab: AppTab = .today

    /// Settings is reached from the Today gear, not from the tab bar.
    var showingSettings = false

    /// The export/import handoff, one level in from Settings. Kept here rather
    /// than as `@State` on the settings screen so "Start over" and a database
    /// replace can close everything from one place.
    var showingBackup = false

    // MARK: Overlays

    /// The product editor sheet — `nil` closed, otherwise create or edit.
    var productEditor: ProductEditorTarget?

    /// The add-stock sheet.
    var addStock: AddStockTarget?

    /// The receipt, shown full-screen after a bill is saved.
    /// Whether Sell is showing the product picker rather than the bill form.
    ///
    /// Screen-local by rights, and here because the shell has to know: the picker
    /// carries its own bottom bar, and the tab bar underneath it would be a
    /// second one stacked on the first. Cleared when Sell goes away, so coming
    /// back lands on the form rather than wherever the last visit ended.
    var pickingProducts = false

    var receipt: Bill?

    /// A bill opened from history. Distinct from `receipt`: that one is a
    /// confirmation of something that just happened, this one is a document
    /// being looked up.
    var billDetail: Bill?

    /// The customer editor sheet — `nil` closed. Carries the customer being
    /// corrected, or `.creating` for a new one.
    var customerEditor: CustomerEditorTarget?

    /// The record-a-payment sheet, for one customer.
    var paymentFor: Customer?

    /// The payment being corrected, if one is. Held beside `paymentFor` rather
    /// than replacing it: the sheet needs the customer either way, because it
    /// shows what will still be owed once the correction is saved.
    var editingPayment: Payment?

    /// The credit-note sheet. Carries the customer it is against, and — when one
    /// is being corrected rather than written — the note itself.
    var creditNoteFor: CreditNoteTarget?

    /// A customer's statement, full screen. Held as a **key** rather than a
    /// `Customer`, because recording a payment while it is open changes every
    /// derived figure on it — the screen has to re-read the customer, not show a
    /// copy taken when it opened.
    var statementFor: String?

    /// A delivery opened from the book, the way a bill is opened from history.
    var purchaseDetail: Purchase?

    /// Set when the add-stock sheet should open on its purchase half.
    var startingPurchase = false

    /// The supplier editor sheet. The customer editor's mirror, kept as its own
    /// pair of fields rather than one editor with a direction on it: the two
    /// sheets say different words and gate on different figures.
    var supplierEditor: SupplierEditorTarget?

    /// The pay-a-supplier sheet.
    var supplierPaymentFor: Supplier?

    /// The money-out twin of `editingPayment`.
    var editingSupplierPayment: SupplierPayment?

    /// The two lists behind Today's banners: everyone who owes the shop, and
    /// everyone the shop owes. Booleans rather than carried values — the sheets
    /// read the whole list off the store, and settling one debtor from inside one
    /// must not close it.
    var showingDebtors = false
    var showingCreditors = false

    /// A supplier's statement, full screen — a key for the same reason
    /// `statementFor` is one, and a separate field so the screen knows which side
    /// of the book it is drawing without being told twice.
    var supplierStatementFor: String?

    // MARK: Intents
    //
    // Named for what the owner is doing, so call sites read as the design does.

    func openNewProduct() {
        productEditor = ProductEditorTarget(product: nil)
    }

    func openNewCustomer() {
        customerEditor = CustomerEditorTarget(customer: nil)
    }

    func openCustomer(_ customer: Customer) {
        customerEditor = CustomerEditorTarget(customer: customer)
    }

    func openStatement(for customer: Customer) {
        statementFor = customer.key
    }

    func openNewSupplier() {
        supplierEditor = SupplierEditorTarget(supplier: nil)
    }

    func openSupplier(_ supplier: Supplier) {
        supplierEditor = SupplierEditorTarget(supplier: supplier)
    }

    func openStatement(forSupplier supplier: Supplier) {
        supplierStatementFor = supplier.key
    }

    func openProduct(_ product: Product) {
        productEditor = ProductEditorTarget(product: product)
    }

    func openAddStock(for product: Product) {
        productEditor = nil
        startingPurchase = false
        addStock = AddStockTarget(product: product)
    }

    /// The Items header's Delivery button, and the Book's empty state.
    ///
    /// No product is named: one is optional on a supplier's bill, and asking for
    /// it first was the app insisting on the answer to a question the paper often
    /// does not have. The sheet offers the catalogue inside itself instead.
    func recordDelivery() {
        startingPurchase = true
        addStock = AddStockTarget(product: nil)
    }

    func openDelivery(for product: Product) {
        startingPurchase = true
        addStock = AddStockTarget(product: product)
    }

    /// Reopens a saved delivery on the sheet it was entered on.
    ///
    /// The product comes from the store rather than from the purchase's snapshot,
    /// because the sheet needs a live one to price against — and it is nil for a
    /// supplier bill that named none, or one whose product has since been deleted.
    func editDelivery(_ purchase: Purchase, product: Product?) {
        purchaseDetail = nil
        startingPurchase = true
        addStock = AddStockTarget(product: product, purchase: purchase)
    }

    func openBill(_ bill: Bill) {
        billDetail = bill
    }

    func startBill() {
        tab = .sell
    }

    func closeOverlays() {
        productEditor = nil
        addStock = nil
        receipt = nil
        billDetail = nil
        showingBackup = false
        customerEditor = nil
        paymentFor = nil
        editingPayment = nil
        creditNoteFor = nil
        statementFor = nil
        supplierEditor = nil
        supplierPaymentFor = nil
        editingSupplierPayment = nil
        supplierStatementFor = nil
        purchaseDetail = nil
        startingPurchase = false
        showingDebtors = false
        showingCreditors = false
    }
}

/// Identifies the supplier editor sheet. `nil` supplier means "New supplier".
struct SupplierEditorTarget: Identifiable {
    let supplier: Supplier?
    var id: String { supplier?.key ?? "new" }
}

/// Identifies the product editor sheet. `nil` product means "New product".
struct ProductEditorTarget: Identifiable {
    let product: Product?
    var id: String { product?.uid.uuidString ?? "new" }
}

/// Identifies the add-stock sheet.
///
/// `product` is nil for a supplier bill that names none — which the sheet can be
/// opened on only by correcting one, since nothing on iOS writes one yet.
/// `purchase` is the delivery being corrected, nil for a new one.
struct AddStockTarget: Identifiable {
    let product: Product?
    var purchase: Purchase? = nil
    var id: String { purchase?.id.uuidString ?? product?.uid.uuidString ?? "new" }
}

/// Identifies the credit-note sheet: who it is against, and the note being
/// corrected where there is one.
///
/// The two travel together because the sheet needs both, and a note with no
/// customer behind it is not a thing this app can draw.
struct CreditNoteTarget: Identifiable {
    let customer: Customer
    var note: CreditNote? = nil
    var id: String { note?.id.uuidString ?? "new-\(customer.key)" }
}

/// Identifies the customer editor sheet. `nil` customer means "New customer".
struct CustomerEditorTarget: Identifiable {
    let customer: Customer?
    var id: String { customer?.key ?? "new" }
}
